README.md
# silverstripe-colorpicker

The ColorPicker Module adds a color-picker input field to the SilverStripe CMS. It makes use of the ColorPicker jQuery
Plugin.

## Requirements

SilverStripe Framework v4+

For a Version that is compatible to SilverStripe 3+, consider using the [3.0 release](https://github.com/tractorcow/silverstripe-colorpicker/releases/tag/3.0.0)

## Installation

Install using composer

```bash
composer require tractorcow/silverstripe-colorpicker ^4@dev
```

## Usage

Here's how you define a DB field to be a color:

```php
private static $db = array(
    'BgColor' => 'TractorCow\Colorpicker\Color'
);
```
    
That's all... scaffolding will take care of creating the appropriate form-field.

If you use `getCMSFields` to create your fields yourself, you might want to do something like this:

```php
public function getCMSFields()
{
    $fields = parent::getCMSFields();

    $fields->addFieldToTab(
    	'Root.Main', 
    	ColorField::create('BgColor', 'Background color')
    );

    return $fields;
}
```

### Tips for using the Color fieldtype in templates

The `TractorCow\Colorpicker\Color` fieldtype provides some helper methods that can be useful in templating. 
Let's consider the above scenario where you have a Field named 'BgColor'. The most common use-case is something like this:

```html
<body style="background-color: #$BgColor;">
...
```

But there's more. You could also use CSS3 `rgba` color definitions with alpha. Example:

```html
<body style="background-color: #$BgColor; background-color: $BgColor.CSSColor(0.5);">
...
```

This will color your background with an alpha value of `0.5` (browsers that don't support rgba, such as IE-8 will fall back to the first background-color definition, that's why it's still in there).

Here's a complete list of the `Color` methods available in templates:

 - `Red` returns the red color component
 - `Green` returns the green color component
 - `Blue` returns the blue color component
 - `CSSColor` returns the color as `rgba`. The alpha value can be specified with the (optional) argument.
 - `Luminance` the luminance of the color as a floating-point value ranging from 0-1
 - `Blend` blends the color with a second background color (defaults to #FFFFFF) with the given opacity. `$BGColor.Blend(0.5, '#000000')` will give the color 50% opacity and put it on top of a black background.
 - `AlteredColorHSV` modifies the current color by the given HSV values. These values are offsets, so you could do something like this: `$BgColor.AlteredColorHSV(0.5, 0, 0)` which will return the color with the opposite hue. All parameters are percentage based and range from `0 - 1`. So doing: `$BgColor.AlteredColorHSV(0, 0, -0.2)` will result in a color with 20% less brightness (absolute, not relative).


CHANGELOG.md
# silverstripe-colorpicker

## 4.0

 - SilverStripe 4.x compatible

## 3.0

 - SilverStripe 3.x compatible.
 - Added custom field-type `Color`

## 1.0

Initial import from http://bummzack.ch/colorpicker/


tests/php/ColorTest.php
<?php

namespace TractorCow\Colorpicker\Tests;

use SilverStripe\Dev\SapphireTest;
use TractorCow\Colorpicker\Color;

class ColorTest extends SapphireTest
{
    /**
     * Test conversion from RGB to HSV
     */
    public function testRGBtoHSV()
    {
        // Red
        $hsv = Color::RGB_TO_HSV(255, 0, 0);
        $this->assertEquals([0, 1, 1], $hsv);

        // Yellow
        $hsv = Color::RGB_TO_HSV(255, 255, 0);
        $this->assertEquals([1 / 360 * 60, 1, 1], $hsv);

        // Teal
        $hsv = Color::RGB_TO_HSV(63.75, 127.5, 127.5);
        $this->assertEquals([.5, .5, .5], $hsv);

        // Black
        $hsv = Color::RGB_TO_HSV(0, 0, 0);
        $this->assertEquals([0, 0, 0], $hsv);

        // Invalid? Should be equal to white
        $hsv = Color::RGB_TO_HSV(400, 330, 890);
        $this->assertEquals([0, 0, 1], $hsv);
    }

    /**
     * Test conversion from HSV to RGB
     */
    public function testHSVtoRGB()
    {
        // Red
        $rgb = Color::HSV_TO_RGB(0, 1, 1);
        $this->assertEquals([255, 0, 0], $rgb);

        // Yellow
        $rgb = Color::HSV_TO_RGB(1 / 360 * 60, 1, 1);
        $this->assertEquals([255, 255, 0], $rgb);

        // Teal
        $rgb = Color::HSV_TO_RGB(.5, .5, .5);
        $this->assertEquals([63.75, 127.5, 127.5], $rgb);

        // Black
        $rgb = Color::HSV_TO_RGB(0, 0, 0);
        $this->assertEquals([0, 0, 0], $rgb);

        // Invalid values. Make sure these are properly clamped. Result should be red
        $rgb = Color::HSV_TO_RGB(-10, 4, 2);
        $this->assertEquals([255, 0, 0], $rgb);
    }

    /**
     * Test HEX to RGB
     */
    public function testHEXtoRGB()
    {
        $rgb = Color::HEX_TO_RGB('#ff0000');
        $this->assertEquals([255, 0, 0], $rgb);

        $rgb = Color::HEX_TO_RGB('#ff8000');
        $this->assertEquals([255, 128, 0], $rgb);

        $rgb = Color::HEX_TO_RGB('#408080');
        $this->assertEquals([64, 128, 128], $rgb);
    }

    /**
     * Test RGB to HEX
     */
    public function testRGBtoHEX()
    {
        $hex = Color::RGB_TO_HEX(255, 0, 0);
        $this->assertEquals('FF0000', $hex);

        $hex = Color::RGB_TO_HEX(255, 128, 0);
        $this->assertEquals('FF8000', $hex);

        $hex = Color::RGB_TO_HEX(64, 128, 128);
        $this->assertEquals('408080', $hex);

        // invalid color values should be clamped
        $hex = Color::RGB_TO_HEX(800, -30, 256);
        $this->assertEquals('FF00FF', $hex);
    }

    /**
     * Test RGB to Luminance
     */
    public function testRGBtoLuminance()
    {
        $l = Color::RGB_TO_LUMINANCE(255, 0, 0);
        $this->assertEquals(.2126, $l);

        $l = Color::RGB_TO_LUMINANCE(0, 255, 0);
        $this->assertEquals(.7152, $l);

        $l = Color::RGB_TO_LUMINANCE(0, 0, 255);
        $this->assertEquals(.0722, $l);

        $l = Color::RGB_TO_LUMINANCE(255, 255, 255);
        $this->assertEquals(1, $l);
    }

    /**
     * Test different color methods
     */
    public function testColorMethods()
    {
        $color = new Color();
        $color->setValue('4080ff');

        // Color-Components
        $this->assertEquals(64, $color->Red());
        $this->assertEquals(128, $color->Green());
        $this->assertEquals(255, $color->Blue());

        // Luminance
        $this->assertEquals(64 / 255 * .2126 + 128 / 255 * .7152 + .0722, $color->Luminance());

        // CSS
        $this->assertEquals('rgba(64,128,255,1.000000)', $color->CSSColor());
        $this->assertEquals('rgba(64,128,255,0.500000)', $color->CSSColor(.5));

        $this->assertEquals('4080ff', $color->forTemplate());
    }

    /**
     * Test color blending method
     */
    public function testBlend()
    {
        $color = new Color();
        $color->setValue('ff8040');

        $mix = $color->Blend(.75);
        // make sure we get a color
        $this->assertInstanceOf(Color::class, $mix);
        $this->assertEquals('FF9F6F', $mix->forTemplate());

        $this->assertEquals('FFDFCF', $color->Blend(.25)->forTemplate());
        $this->assertEquals('402010', $color->Blend(.25, '#000000')->forTemplate());
        $this->assertEquals('80809F', $color->Blend(.50, '0080ff')->forTemplate());

        // Blending with 1 or 0 opacity
        $this->assertEquals('FFFFFF', $color->Blend(0)->forTemplate());
        $this->assertEquals('FF8040', $color->Blend(1)->forTemplate());

        // Chain methods
        $this->assertEquals('C0CF78', $color->Blend(.50)->Blend(.75, '#00ff00')->forTemplate());
    }

    /**
     * Test color HSV shifting method
     */
    public function testHsvAlteration()
    {
        $color = new Color();
        $color->setValue('80FF40');

        $altered = $color->AlteredColorHSV(.5, 0, 0);
        // make sure we get a color
        $this->assertInstanceOf(Color::class, $altered);
        $this->assertEquals('BE40FF', $altered->forTemplate());
        $this->assertEquals('A900FF', $color->AlteredColorHSV(.5, .5, 0)->forTemplate());
        $this->assertEquals('557F3F', $color->AlteredColorHSV(0, -.25, -.5)->forTemplate());
        // No change doesn't change the color value
        $this->assertEquals('80FF40', $color->AlteredColorHSV(0, 0, 0)->forTemplate());
    }
}


src/Forms/ColorField.php
<?php

namespace TractorCow\Colorpicker\Forms;

use TractorCow\Colorpicker\Color;
use SilverStripe\Forms\Form;
use SilverStripe\View\Requirements;
use SilverStripe\Forms\TextField;

/**
 * Color field
 */
class ColorField extends TextField
{
    /**
     * Returns a color field.
     *
     * @param string $name
     * @param null|string $title
     * @param string $value
     * @param null|Form $form
     */
    public function __construct($name, $title = null, $value = '', $form = null)
    {
        parent::__construct($name, $title, $value, 16, $form);
    }

    public function Field($properties = [])
    {
        $this->addExtraClass('colorfield');
        Requirements::javascript(COLORPICKER_DIR . '/client/javascript/colorpicker.js');
        Requirements::javascript(COLORPICKER_DIR . '/client/javascript/colorfield.js');
        Requirements::css(COLORPICKER_DIR . '/client/css/colorpicker.css');

        $this->setAttribute('style', sprintf(
            'background-image: none; background-color: %s; %s',
            ($this->value ? '#' . $this->value : '#ffffff'), $this->getTextColorStyle()
        ));

        return parent::Field($properties);
    }

    /**
     * Override the type to get the proper class name on the field
     * "text" is needed here to render the form field as a normal text-field
     * @see FormField::Type()
     */
    public function Type()
    {
        return 'text';
    }

    /**
     * Ensure the color is a valid hexadecimal color
     * @see FormField::validate()
     *
     * @param \SilverStripe\Forms\Validator $validator
     * @return bool whether or not the field is valid
     */
    public function validate($validator)
    {
        if (!empty($this->value) && !preg_match('/^[A-f0-9]{6,8}$/', $this->value)) {
            $validator->validationError(
                $this->name,
                _t('ColorField.VALIDCOLORFORMAT', 'Please enter a valid color in hexadecimal format.'),
                'validation',
                false
            );
            return false;
        }
        return true;
    }

    /**
     * Calculate the text color style so that it's visible on the background
     * @return string
     */
    protected function getTextColorStyle()
    {
        // change alpha component depending on disabled state
        $a = $this->isDisabled() ? '0.5' : '1.0';
        if ($this->value) {
            list($R, $G, $B) = Color::HEX_TO_RGB($this->value);
            $luminance = Color::RGB_TO_LUMINANCE($R, $G, $B);
            // return color as hex and as rgba values (hex is fallback for IE-8)
            return ($luminance > 0.5) ?
                'color: #000; color: rgba(0, 0, 0, '. $a .');' :
                'color: #fff; color: rgba(255, 255, 255, '. $a .');';
        } else {
            return 'color: #000; color: rgba(0, 0, 0, '. $a .');';
        }
    }
}


client/javascript/colorfield.js
(function($) {
	$.entwine('ss', function($) {
		$('input.colorfield').entwine({
			onmatch: function(e) {
				var self=$(this);
				//Calculate luminance (Photometric/digital ITU-R)
				var calcLuminance = function(rgb){
					return Math.min(1, Math.max(0, 0.2126 * (rgb.r / 255) + 0.7152 * (rgb.g / 255) + 0.0722 * (rgb.b / 255)));
				};
				var picker=$(this).ColorPicker({
					onSubmit: function(hsb, hex, rgb) {
						var col=(calcLuminance(rgb) > 0.5 ? '#000000':'#ffffff');
						self.val(hex).css({color:col, backgroundColor:'#'+hex});
					},
					onBeforeShow: function() {
						$(this).ColorPickerSetColor(this.value);
					},
					onChange: function (hsb, hex, rgb) {
						var col=(calcLuminance(rgb)  > 0.5 ? '#000000':'#ffffff');
						self.val(hex).css({color:col, backgroundColor:'#'+hex});
					}
				});
			},
			onkeyup: function() {
				$(this).ColorPickerSetColor($(this).val());
			}
		});
	});
})(jQuery);


src/Color.php
<?php

namespace TractorCow\Colorpicker;

use SilverStripe\ORM\FieldType\DBVarchar;
use TractorCow\Colorpicker\Forms\ColorField;

/**
 * Color field-type
 * @author bummzack
 */
class Color extends DBVarchar
{
    private static $casting = [
        'Luminance' => 'Float',
        'AlteredColorHSV' => 'TractorCow\Colorpicker\Color'
    ];

    /**
     * Helper function to convert RGB to HSV
     * @param int $R red channel, 0-255
     * @param int $G green channel, 0-255
     * @param int $B blue channel, 0-255
     * @return array containing 3 values, H,S,V 0-1
    */
    public static function RGB_TO_HSV($R, $G, $B)  // RGB Values:Number 0-255
    {
        $HSV = [];

        $var_R = self::clamp($R / 255);
        $var_G = self::clamp($G / 255);
        $var_B = self::clamp($B / 255);

        $var_Min = min($var_R, $var_G, $var_B);
        $var_Max = max($var_R, $var_G, $var_B);
        $del_Max = $var_Max - $var_Min;

        $V = $var_Max;
        $H = 0;

        if ($del_Max == 0) {
            $H = 0;
            $S = 0;
        } else {
            $S = $del_Max / $var_Max;

            $del_R = ((($var_Max - $var_R) / 6) + ($del_Max / 2)) / $del_Max;
            $del_G = ((($var_Max - $var_G) / 6) + ($del_Max / 2)) / $del_Max;
            $del_B = ((($var_Max - $var_B) / 6) + ($del_Max / 2)) / $del_Max;

            if ($var_R == $var_Max) {
                $H = $del_B - $del_G;
            } elseif ($var_G == $var_Max) {
                $H = (1 / 3) + $del_R - $del_B;
            } elseif ($var_B == $var_Max) {
                $H = (2 / 3) + $del_G - $del_R;
            }

            if ($H<0) {
                $H++;
            }
            if ($H>1) {
                $H--;
            }
        }

        $HSV[] = $H;
        $HSV[] = $S;
        $HSV[] = $V;

        return $HSV;
    }

    /**
     * Helper function to convert HSV to RGB
     * @param float $H hue 0-1
     * @param float $S saturation 0-1
     * @param float $V brightness 0-1
     * @return array containing 3 values in the range from 0-255, R,G,B
     */
    public static function HSV_TO_RGB($H, $S, $V)
    {
        $H = self::clamp($H * 6, 0, 6);
        $S = self::clamp($S);
        $V = self::clamp($V);
        $I = floor($H);
        $F = $H - $I;
        $M = $V * (1 - $S);
        $N = $V * (1 - $S * $F);
        $K = $V * (1 - $S * (1 - $F));
        $R = $G = $B = 0;
        switch ($I) {
            case 0:
                list($R, $G, $B) = [$V,$K,$M];
                break;
            case 1:
                list($R, $G, $B) = [$N,$V,$M];
                break;
            case 2:
                list($R, $G, $B) = [$M,$V,$K];
                break;
            case 3:
                list($R, $G, $B) = [$M,$N,$V];
                break;
            case 4:
                list($R, $G, $B) = [$K,$M,$V];
                break;
            case 5:
            case 6: //for when $H=1 is given
                list($R, $G, $B) = [$V,$M,$N];
                break;
        }
        return [$R * 255.0, $G * 255.0, $B * 255.0];
    }

    /**
     * Convert a hex string to separate R,G,B values
     * @param string $hex
     * @return array containing 3 integers (0-255) R,G,B
     */
    public static function HEX_TO_RGB($hex)
    {
        $RGB = [];

        $color = intval(ltrim($hex, '#'), 16);
        $r = ($color >> 16) & 0xff;
        $g = ($color >> 8) & 0xff;
        $b = $color & 0xff;

        $RGB[] = $r;
        $RGB[] = $g;
        $RGB[] = $b;

        return $RGB;
    }

    /**
     * Convert R,G,B to hex
     * @param int $R
     * @param int $G
     * @param int $B
     * @return string
     */
    public static function RGB_TO_HEX($R, $G, $B)
    {
        $R = self::clamp($R, 0, 255);
        $G = self::clamp($G, 0, 255);
        $B = self::clamp($B, 0, 255);
        return sprintf("%06X", ($R << 16) | ($G << 8) | $B);
    }

    /**
     * Calculate luminance (Photometric/digital ITU-R)
     * @param int $R
     * @param int $G
     * @param int $B
     * @return number 0-1
     */
    public static function RGB_TO_LUMINANCE($R, $G, $B)
    {
        return self::clamp(0.2126 * ($R / 255) + 0.7152 * ($G / 255) + 0.0722 * ($B / 255));
    }

    public function __construct($name = null, $options = [])
    {
        parent::__construct($name, 16, $options);
    }

    public function scaffoldFormField($title = null, $params = null)
    {
        $field = new ColorField($this->name, $title);
        return $field;
    }

    /**
     * Get the red component of this color
     * @return int red-component 0-255
     */
    public function Red()
    {
        list($R, $G, $B) = self::HEX_TO_RGB($this->value);
        return $R;
    }

    /**
     * Get the green component of this color
     * @return int green-component 0-255
     */
    public function Green()
    {
        list($R, $G, $B) = self::HEX_TO_RGB($this->value);
        return $G;
    }

    /**
     * Get the blue component of this color
     * @return int blue-component 0-255
     */
    public function Blue()
    {
        list($R, $G, $B) = self::HEX_TO_RGB($this->value);
        return $B;
    }

    /**
     * Get the color as CSS3 color definition with optional alpha value.
     * Will return "rgba(RED, GREEN, BLUE, OPACITY)"
     * @param number $opacity opacity value from 0-1
     * @return string css3 color definition
     */
    public function CSSColor($opacity = 1)
    {
        list($R, $G, $B) = self::HEX_TO_RGB($this->value);
        $A = self::clamp($opacity);
        return sprintf('rgba(%d,%d,%d,%f)', $R, $G, $B, $A);
    }

    /**
     * Return the luminance of the color
     * @return float the luminance 0-1
     */
    public function Luminance()
    {
        list($R, $G, $B) = self::HEX_TO_RGB($this->value);
        return self::RGB_TO_LUMINANCE($R, $G, $B);
    }

    /**
     * Change the color by the given HSV values and return a new color
     * @param float $hChange hue change
     * @param float $sChange saturation change
     * @param float $vChange brightness change
     * @return Color the new color
     */
    public function AlteredColorHSV($hChange, $sChange, $vChange)
    {
        list($R, $G, $B) = self::HEX_TO_RGB($this->value);
        list($H, $S, $V) = self::RGB_TO_HSV($R, $G, $B);
        list($R, $G, $B) = self::HSV_TO_RGB(
            fmod($H + $hChange + 1, 1),
            self::clamp($S + $sChange),
            self::clamp($V + $vChange)
        );

        $color = new Color();
        $color->setValue(self::RGB_TO_HEX($R, $G, $B));
        return $color;
    }

    /**
     * Blend the color with a background color, with the given opacity level
     * @param float $opacity Opacity level of the current color (between 0 - 1)
     * @param string $background The background color
     * @return Color the new color
     */
    public function Blend($opacity, $background = 'FFFFFF')
    {
        list($R, $G, $B) = self::HEX_TO_RGB($this->value);
        list($bgR, $bgG, $bgB) = self::HEX_TO_RGB($background);

        $transparency = self::clamp(1 - $opacity);

        $R += intval(($bgR - $R) * $transparency);
        $G += intval(($bgG - $G) * $transparency);
        $B += intval(($bgB - $B) * $transparency);

        $color = new Color();
        $color->setValue(self::RGB_TO_HEX($R, $G, $B));
        return $color;
    }

    /**
     * Clamp value to min/max
     * @param float|int $val
     * @param float|int $min, defaults to 0
     * @param float|int $max, defaults to 1
     * @return float|int value clamped to min/max.
     */
    private static function clamp($val, $min = 0, $max = 1)
    {
        return min($max, max($min, $val));
    }
}
